src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -0,0 +1,100 @@
+
+#include <Interpreters/TableJoin.h>
+
+#include <Common/logger_useful.h>
+#include <Core/Joins.h>
+#include <ranges>
+
+namespace DB::QueryPlanOptimizations
+{
+
+static std::optional<UInt64> estimateReadRowsCount(QueryPlan::Node & node)
+{
+    IQueryPlanStep * step = node.step.get();
+    if (const auto * reading = typeid_cast<const ReadFromMergeTree *>(step))
+    {
+        if (auto analyzed_result = reading->getAnalyzedResult())
+            return analyzed_result->selected_rows;
+        if (auto analyzed_result = reading->selectRangesToRead())
+            return analyzed_result->selected_rows;
+        return {};
+    }
+
+    if (const auto * reading = typeid_cast<const ReadFromMemoryStorageStep *>(step))
+        return reading->getStorage()->totalRows(Settings{});
+
+    if (node.children.size() != 1)
+        return {};
+
+    if (typeid_cast<ExpressionStep *>(step) || typeid_cast<FilterStep *>(step))
+        return estimateReadRowsCount(*node.children.front());
+
+    return {};
+}
+
+void optimizeJoin(QueryPlan::Node & node, QueryPlan::Nodes &)
+{
+    auto * join_step = typeid_cast<JoinStep *>(node.step.get());
+    if (!join_step || node.children.size() != 2)
+        return;
+
+    const auto & join = join_step->getJoin();
+    if (join->pipelineType() != JoinPipelineType::FillRightFirst || !join->isCloneSupported() || typeid_cast<const HashJoin *>(join.get()))
+        return;
+
+    const auto & table_join = join->getTableJoin();
+    auto kind = table_join.kind();
+    if (table_join.hasUsing()
+     || table_join.strictness() != JoinStrictness::All
+     || (kind != JoinKind::Inner && kind != JoinKind::Left
+      && kind != JoinKind::Right && kind != JoinKind::Full))
+        return;
+
+    bool need_swap = false;
+    if (join_step->inner_table_selection_mode == JoinInnerTableSelectionMode::Auto)
+    {
+        auto lhs_extimation = estimateReadRowsCount(*node.children[0]);
+        auto rhs_extimation = estimateReadRowsCount(*node.children[1]);
+        LOG_TRACE(getLogger("optimizeJoin"), "Left table estimation: {}, right table estimation: {}",
+            lhs_extimation.transform(toString<UInt64>).value_or("unknown"),
+            rhs_extimation.transform(toString<UInt64>).value_or("unknown"));
+
+        if (lhs_extimation && rhs_extimation && *lhs_extimation < *rhs_extimation)
+            need_swap = true;
+    }
+    else if (join_step->inner_table_selection_mode == JoinInnerTableSelectionMode::Left)
+    {
+        need_swap = true;
+    }
+
+    if (!need_swap)
+        return;
+
+    const auto & streams = join_step->getInputStreams();
+    if (streams.size() != 2)
+        return;
+
+    const auto & left_stream_input_header = streams.front().header;
+    const auto & right_stream_input_header = streams.back().header;
+    join_step->swap_streams = true;
+
+    auto updated_table_join = std::make_shared<TableJoin>(table_join);
+    updated_table_join->swapSides();
```

Comment:
it seems it could happen that different nodes could make different optimisation decisions here in a distributed query. afaiu we change output header too. is it actually a problem and should we then maybe take some actions to preserve the header?

Reply:
Implemented transformations inside JoinStep, so it keeps column order unchanged